src/ws/metrics.rs

```diff
@@ -64,24 +148,95 @@ impl Procedure {
     }
 }
 
-pub async fn record_procedure_call(
+/// Records a procedure call. This increments the counter of procedure calls
+/// based on the response code and the specific procedure.
+async fn record_procedure_call(
+    metrics: Arc<Metrics>,
+    code: Option<WsServiceError>,
+    procedure: Procedure,
+) {
+    let code = map_error_code(code);
+    metrics
+        .procedure_call_total_collector
+        .with_label_values(&[code, procedure.as_str()])
+        .inc();
+}
+
+/// Increments the count of connected clients.
+pub async fn increment_connected_clients(metrics: Arc<Metrics>) {
+    metrics.connected_clients_total_collector.inc();
+}
+
+/// Decrements the count of connected clients.
+pub async fn decrement_connected_clients(metrics: Arc<Metrics>) {
+    metrics.connected_clients_total_collector.dec();
+}
+
+/// Records updates sent on subscription. This increments the counter of updates sent
+/// on subscription based on the event type.
+pub async fn record_updates_sent(metrics: Arc<Metrics>, event: FriendshipEvent) {
+    metrics
+        .updates_sent_on_subscription_total_collector
+        .with_label_values(&[event.as_str()])
+        .inc();
+}
+
+/// Records the size of a procedure call. This adds the size of the procedure call to the
+/// histogram for the specified procedure.
+pub async fn record_procedure_call_size<T: prost::Message>(
+    metrics: Arc<Metrics>,
+    procedure: Procedure,
+    msg: &T,
+) {
+    let size = calculate_message_size(msg);
+    metrics
+        .procedure_call_size_bytes_histogram_collector
+        .with_label_values(&[procedure.as_str()])
+        .observe(size as f64);
+}
+
+/// Records the duration of a procedure call. This adds the duration of the procedure call to the
+/// histogram for the specified procedure and response code.
+async fn record_procedure_call_duration(
+    metrics: Arc<Metrics>,
+    code: Option<WsServiceError>,
+    procedure: Procedure,
+    start_time: Instant,
+) {
+    let code = map_error_code(code);
+    let duration = Instant::now().duration_since(start_time).as_secs_f64();
+    metrics
+        .procedure_call_duration_seconds_histogram_collector
+        .with_label_values(&[code, procedure.as_str()])
+        .observe(duration);
+}
+
+/// Records a procedure call and its duration.
+pub async fn record_procedure_call_and_duration(
```

Comment:
what do you think of adding another responsibility to this method so it records the size too?

Reply:
I think it depends on the above comment, but I'll definitely change it if we're reporting the sum of both request and response size.